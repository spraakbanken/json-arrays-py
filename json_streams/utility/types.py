import os
import sys
import typing

# pylint: disable=unsubscriptable-object

if sys.version_info >= (3, 9):
    Pathlike = typing.TypeVar(
        "Pathlike", str, bytes, os.PathLike[str], os.PathLike[bytes]
    )
else:
    Pathlike = typing.TypeVar("Pathlike", str, bytes, os.PathLike)
