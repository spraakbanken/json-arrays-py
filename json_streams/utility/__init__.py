"""Util functions.
"""
from enum import Enum
import os
from typing import Any, Generator, IO, Union

# pylint: disable=unsubscriptable-object


def get_name_of_file(fp: IO) -> str:
    if hasattr(fp, "name"):
        return fp.name
    return ""


def is_jsonl(name: str) -> bool:
    """Test if a filename is json lines.
    Also returns True for '<stdin>' and '<stdout>'.

    Arguments:
        name {str} -- name to test

    Returns:
        {bool} -- True if this name is jsonl.
    """
    if name in ["<stdin>", "<stdout>"]:
        return True
    _, suffix = os.path.splitext(name)
    # print('suffix = {suffix}'.format(suffix=suffix))
    return suffix in [".jsonl"]


def to_bytes(s: Union[str, bytes, bytearray]) -> Union[bytes, bytearray]:
    """Convert str to bytes, otherwise pass through s

    Args:
        s (Union[str, bytes, bytearray]): the argument to assure bytes

    Returns:
        Union[bytes, bytearray]: the possible converted argument
    """
    if isinstance(s, (bytes, bytearray)):
        return s
    return s.encode("utf-8")


class Sink:
    def __init__(self, sink: Generator[None, Any, None]):
        self.sink = sink

    def __enter__(self):
        next(self.sink)
        return self.sink

    def __exit__(self, exc_type, exc_value, exc_traceback):
        self.sink.close()


class JsonFormat(str, Enum):
    JSON = "json"
    JSON_LINES = "jsonl"
