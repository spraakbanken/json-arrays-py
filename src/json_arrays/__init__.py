"""Handle JSON or JSON_LINES lazily."""
import sys
import typing
from typing import BinaryIO, Iterable, Optional

from json_arrays import _types, encoders, json_iter, jsonl_iter, jsonlib, utility

__all__ = ["encoders", "json_iter", "jsonl_iter", "jsonlib", "utility"]


# pylint: disable=unsubscriptable-object
def choose_iter(name, json_format: Optional[utility.JsonFormat]):
    if json_format == utility.JsonFormat.JSON_LINES or utility.is_jsonl(name):
        return jsonl_iter
    return json_iter


def load(fp: BinaryIO, *, json_format: Optional[utility.JsonFormat] = None) -> Iterable:
    """Dump an iterable to file-like object.

    Examples:
        Loading iterable in JSON format:
        >>> import io
        >>> import json_arrays
        >>> in_json = io.BytesIO(b'[{"a": 1}, {"a": 2}]')
        >>> iter = json_arrays.load(in_json)
        >>> list(iter)
        [{'a': 1}, {'a': 2}]

        Loading iterable in JSON_LINES format:
        >>> import io
        >>> import json_arrays
        >>> in_jsonl = io.BytesIO(b'{"a": 1}\\n{"a": 2}\\n')
        >>> iter = json_arrays.load(in_jsonl, json_format="jsonl")
        >>> list(iter)
        [{'a': 1}, {'a': 2}]
    """
    _iter = choose_iter(utility.get_name_of_file(fp), json_format)

    yield from _iter.load(fp)


def load_from_file(
    file_name: typing.Optional[_types.Pathlike],
    *,
    json_format: Optional[utility.JsonFormat] = None,
    file_mode: str = "rb",
    use_stdin_as_default: bool = False,
    **kwargs,
) -> Iterable:
    """Lazily load from given file_name.

    Reads from stdin if `use_stdin_as_default` is set and file_name is falsy.

    Parameters:
        file_name : Path
            name of file to load from
        json_format : utility.JsonFormat, optional, default=None
            Explicit set format of json file
        file_mode : str, optional, default="rb"
            mode to open the file in.
        use_stdin_as_default : bool, optional, default=False
            reads from stdin if file_name is None

    Yields:
        JSON value
            lazily loaded from file
    """
    if file_name is not None:
        _iter = choose_iter(file_name, json_format)

        yield from _iter.load_from_file(file_name, file_mode=file_mode, **kwargs)
    elif use_stdin_as_default:
        yield from jsonl_iter.load(sys.stdin.buffer, **kwargs)
    else:
        raise ValueError("You must give a FILENAME or USE_STDIN_AS_DEFAULT=`True`")


def dump(in_iter_, fp: BinaryIO, *, json_format: Optional[utility.JsonFormat] = None):
    """Dump an iterable to BinaryIO-object.

    Examples:
        Dumping iterable in JSON format:
        >>> import io
        >>> import json_arrays
        >>> data = [{"a": 1}, {"a": 2}]
        >>> out_json = io.BytesIO()
        >>> json_arrays.dump(data, out_json)
        >>> out_json.getvalue()
        b'[{"a":1},{"a":2}]'

        Dumping iterable in JSON_LINES format:
        >>> import io
        >>> import json_arrays
        >>> data = [{"a": 1}, {"a": 2}]
        >>> out_jsonl = io.BytesIO()
        >>> json_arrays.dump(data, out_jsonl, json_format="jsonl")
        >>> out_jsonl.getvalue()
        b'{"a":1}\\n{"a":2}\\n'
    """
    _iter = choose_iter(utility.get_name_of_file(fp), json_format)

    _iter.dump(in_iter_, fp)


def dump_to_file(
    in_iter_,
    file_name: typing.Optional[_types.Pathlike],
    *,
    json_format: Optional[utility.JsonFormat] = None,
    file_mode: str = "wb",
    use_stdout_as_default: bool = False,
    use_stderr_as_default: bool = False,
    **kwargs,
):
    """Open file and dump json to it.

    Parameters:
        in_iter_ : Iterable
            The data to dump.
        file_name : Path
            The path to write to
        json_format : utility.JsonFormat, optional, default=None
            the json format to write in.
        file_mode : str, optional, default="wb"
            the mode to open the file in.
        use_stdout_as_default : bool, optional, default=False
            use stdout if file_name is empty.
        use_stderr_as_default : bool, optional, default=False
            use stderr if file_name is empty.
    """
    if file_name is not None:
        _iter = choose_iter(file_name, json_format)

        _iter.dump_to_file(in_iter_, file_name, file_mode=file_mode, **kwargs)

    elif use_stdout_as_default:
        jsonl_iter.dump(in_iter_, sys.stdout.buffer, **kwargs)
    elif use_stderr_as_default:
        jsonl_iter.dump(in_iter_, sys.stderr.buffer, **kwargs)
    else:
        raise ValueError(
            "You must give a FILENAME or USE_STDOUT_AS_DEFAULT=`True` or USE_STDERR_AS_DEFAULT=`True`"  # noqa: E501
        )


def sink(fp: BinaryIO, *, json_format: Optional[utility.JsonFormat] = None):
    _iter = choose_iter(utility.get_name_of_file(fp), json_format)

    return _iter.sink(fp)


def sink_from_file(
    file_name: typing.Optional[_types.Pathlike],
    *,
    json_format: Optional[utility.JsonFormat] = None,
    file_mode: str = "wb",
    use_stdout_as_default: bool = False,
    use_stderr_as_default: bool = False,
):
    """Open file and use it as json sink.

    Args:
        file_name : Path
            the path to write to
        json_format : utility.JsonFormat, optional, default=None
            the json format to write in.
        file_mode : str, optional, default="wb"
            the mode to open the file in.
        use_stdout_as_default : bool, optional, default=False
            use stdout if file_name is empty.
        use_stderr_as_default : bool, optional, default=False
            use stderr if file_name is empty.
    """
    if not file_name:
        if use_stdout_as_default:
            return jsonl_iter.sink(sys.stdout.buffer)
        if use_stderr_as_default:
            return jsonl_iter.sink(sys.stderr.buffer)
        raise ValueError(
            "You must give a FILENAME or USE_STDOUT_AS_DEFAULT=`True` or USE_STDERR_AS_DEFAULT=`True`"  # noqa: E501
        )

    _iter = choose_iter(file_name, json_format)

    return _iter.sink_from_file(file_name, file_mode=file_mode)
