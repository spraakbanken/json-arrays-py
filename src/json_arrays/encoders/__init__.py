"""Encoders

`encode_decimal`
>>> from decimal import Decimal
>>> from json_arrays import jsonlib, encoders
>>> jsonlib.dumps([10.20, "10.20", Decimal("10.20")], default=encoders.encode_decimal)
b'[10.2,"10.20",10.2]'
"""

from .decimal_encoders import encode_decimal

__all__ = ["encode_decimal"]
