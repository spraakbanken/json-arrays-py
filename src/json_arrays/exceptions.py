"""Exceptions for json_arrays."""


class JsonStreamsException(Exception):
    """
    Base exception for json_arrays
    """
