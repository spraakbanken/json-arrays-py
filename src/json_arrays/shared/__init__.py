from .writer import JsonArrayWriter

__all__ = ["JsonArrayWriter"]
