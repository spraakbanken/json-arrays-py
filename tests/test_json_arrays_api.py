import gzip
import io
import json
import tempfile
from pathlib import Path
from typing import Optional
from unittest.mock import patch

import json_arrays
import pytest
from json_arrays import files
from json_arrays.utility import JsonFormat


@pytest.mark.parametrize(
    "file_name",
    [
        "tests/data/array.json",
        "tests/data/array.jsonl",
        "tests/data/array.ndjson",
        "tests/data/array.json.gz",
        "tests/data/array.jsonl.gz",
        "tests/data/array.ndjson.gz",
    ],
)
def test_load_from_file(file_name: str, snapshot_json):
    assert list(json_arrays.load_from_file(file_name)) == snapshot_json


@pytest.mark.parametrize(
    "file_name",
    [
        "tests/data/array.json",
        "tests/data/array.jsonl",
        "tests/data/array.ndjson",
    ],
)
def test_load(file_name: str, snapshot_json) -> None:
    with open(file_name, mode="rb") as fp:
        data_loaded = list(json_arrays.load(fp))

    assert data_loaded == snapshot_json


@pytest.mark.parametrize(
    "file_name, json_format",
    [
        ("tests/data/array.json", JsonFormat.JSON),
        ("tests/data/array.jsonl", None),
        ("tests/data/array.jsonl", JsonFormat.JSON_LINES),
    ],
)
def test_load_from_file_stdin(
    file_name: str, json_format: Optional[json_arrays.JsonFormat], snapshot_json
) -> None:
    buffer = io.BytesIO(Path(file_name).read_bytes())
    stdin_patcher = patch("sys.stdin", buffer=buffer)
    with stdin_patcher.start():
        data_loaded = list(
            json_arrays.load_from_file(
                file_name=None, json_format=json_format, use_stdin_as_default=True
            )
        )

    assert data_loaded == snapshot_json


@pytest.mark.parametrize(
    "json_format", [json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_dump(json_format: json_arrays.JsonFormat, snapshot, array_of_dicts: list[dict]) -> None:
    with tempfile.TemporaryFile() as fp:
        json_arrays.dump(array_of_dicts, fp, json_format=json_format)
        fp.seek(0)
        data_written = fp.read()

    assert data_written == snapshot


@pytest.mark.parametrize(
    "json_format", [json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_dump_dict(json_format: json_arrays.JsonFormat, snapshot, data_dict: dict) -> None:
    with tempfile.TemporaryFile() as fp:
        json_arrays.dump(data_dict, fp, json_format=json_format)
        fp.seek(0)
        data_written = fp.read()

    assert data_written == snapshot


@pytest.mark.parametrize(
    "json_format", [json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_dump_int(json_format: json_arrays.JsonFormat, snapshot) -> None:
    with tempfile.TemporaryFile() as fp:
        json_arrays.dump(1234, fp, json_format=json_format)
        fp.seek(0)
        data_written = fp.read()

    assert data_written == snapshot


@pytest.mark.parametrize(
    "json_format", [json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_dump_str(json_format: json_arrays.JsonFormat, snapshot) -> None:
    with tempfile.TemporaryFile() as fp:
        json_arrays.dump("just an ordinary\n string", fp, json_format=json_format)
        fp.seek(0)
        data_written = fp.read()

    assert data_written == snapshot


@pytest.mark.parametrize(
    "file_name",
    [
        "tests/data/gen/api_array.json",
        "tests/data/gen/api_array.jsonl",
        "tests/data/gen/api_array.ndjson",
        "tests/data/gen/api_array.json.gz",
        "tests/data/gen/api_array.jsonl.gz",
        "tests/data/gen/api_array.ndjson.gz",
    ],
)
def test_dump_to_file(file_name: str, snapshot, array_of_dicts: list[dict]) -> None:
    json_arrays.dump_to_file(array_of_dicts, file_name)

    with files.BinaryFileRead(file_name).file as fp:
        bytes_written = fp.read()
    assert bytes_written == snapshot


@pytest.mark.parametrize(
    "json_format", [None, json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_dump_to_file_stdout(
    json_format: Optional[json_arrays.JsonFormat], snapshot, array_of_dicts: list[dict]
) -> None:
    buffer = io.BytesIO()
    stdout_patcher = patch("sys.stdout", buffer=buffer)
    with stdout_patcher.start():
        json_arrays.dump_to_file(
            array_of_dicts, file_name=None, json_format=json_format, use_stdout_as_default=True
        )

        data_written = buffer.getvalue()

    assert data_written == snapshot


@pytest.mark.parametrize(
    "json_format", [None, json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_dump_to_file_stderr(
    json_format: Optional[json_arrays.JsonFormat], snapshot, array_of_dicts: list[dict]
) -> None:
    buffer = io.BytesIO()
    stderr_patcher = patch("sys.stderr", buffer=buffer)
    with stderr_patcher.start():
        json_arrays.dump_to_file(
            array_of_dicts, file_name=None, json_format=json_format, use_stderr_as_default=True
        )

        data_written = buffer.getvalue()

    assert data_written == snapshot


@pytest.mark.parametrize("file_suffix", [".json", ".jsonl", ".ndjson"])
def test_sink(file_suffix: str, snapshot, array_of_dicts: list[dict]) -> None:
    with tempfile.NamedTemporaryFile(suffix=file_suffix) as fp:
        with json_arrays.sink(fp) as sink:  # type: ignore[arg-type]
            for obj in array_of_dicts:
                sink.send(obj)
        fp.seek(0)
        data_written = fp.read()

    assert data_written == snapshot


@pytest.mark.parametrize("file_suffix", [".json", ".jsonl", ".ndjson"])
def test_sink_dict(file_suffix: str, snapshot) -> None:
    with tempfile.NamedTemporaryFile(suffix=file_suffix) as fp:
        with json_arrays.sink(fp) as sink:  # type: ignore[arg-type]
            sink.send({"a": "b"})
        fp.seek(0)
        data_written = fp.read()

    assert data_written == snapshot


@pytest.mark.parametrize(
    "file_name",
    [
        "tests/data/gen/api_array.json",
        "tests/data/gen/api_array.jsonl",
        "tests/data/gen/api_array.ndjson",
        "tests/data/gen/api_array.json.gz",
        "tests/data/gen/api_array.jsonl.gz",
        "tests/data/gen/api_array.ndjson.gz",
    ],
)
def test_sink_from_file(file_name: str, snapshot, array_of_dicts: list[dict]) -> None:
    with json_arrays.sink_from_file(file_name) as sink:
        for obj in array_of_dicts:
            sink.send(obj)

    with files.BinaryFileRead(file_name).file as fp:
        bytes_written = fp.read()
    assert bytes_written == snapshot


@pytest.mark.parametrize(
    "json_format", [None, json_arrays.JsonFormat.JSON, json_arrays.JsonFormat.JSON_LINES]
)
def test_sink_from_file_stdout(
    json_format: Optional[json_arrays.JsonFormat], snapshot, array_of_dicts: list[dict]
) -> None:
    buffer = io.BytesIO()
    stdout_patcher = patch("sys.stdout", buffer=buffer)
    with stdout_patcher.start():
        with json_arrays.sink_from_file(
            file_name=None, json_format=json_format, use_stdout_as_default=True
        ) as sink:
            for obj in array_of_dicts:
                sink.send(obj)

        data_written = buffer.getvalue()

    assert data_written == snapshot


def test_load_from_file_fails_without_file_name(snapshot) -> None:
    try:
        for _ in json_arrays.load_from_file(None, use_stdin_as_default=False):
            pass
    except ValueError as exc:
        assert str(exc) == snapshot


def test_dump_to_file_fails_without_file_name(snapshot) -> None:
    try:
        json_arrays.dump_to_file(
            [], None, use_stdout_as_default=False, use_stderr_as_default=False
        )
    except ValueError as exc:
        assert str(exc) == snapshot


def test_sink_from_file_fails_without_file_name(snapshot) -> None:
    try:
        json_arrays.sink_from_file(
            None, use_stdout_as_default=False, use_stderr_as_default=False
        )
    except ValueError as exc:
        assert str(exc) == snapshot


def test_json_gzip_dump_fp(array_of_dicts: list[dict], snapshot_json):
    filename = Path("tests/data/gen/json_gzip_dump_fp.json.gz")

    with open(filename, "wb") as fp:
        json_arrays.dump(array_of_dicts, fp)

    with gzip.open(filename) as fp:  # type: ignore
        loaded_entries = json.load(fp)

    assert loaded_entries == snapshot_json


def test_json_gzip_load_fp(array_of_dicts: list[dict], snapshot_json):
    filename = Path("tests/data/gen/json_gzip_load_fp.json.gz")

    with gzip.open(filename, "wt") as fp:
        json.dump(array_of_dicts, fp)

    with open(filename, "rb") as fp:  # type: ignore
        loaded_entries = list(json_arrays.load(fp))  # type: ignore

    assert loaded_entries == snapshot_json
