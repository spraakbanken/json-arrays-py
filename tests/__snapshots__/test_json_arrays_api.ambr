# serializer version: 1
# name: test_dump[json]
  b'[{"a":"a1"},{"b":["b1","b2"]},{"c":{"c1":"c11"}}]'
# ---
# name: test_dump[jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_dict[json]
  b'{"a":[{"b":2,"c":"hi","d":[1,2]}],"b":[1,2]}'
# ---
# name: test_dump_dict[jsonl]
  b'{"a":[{"b":2,"c":"hi","d":[1,2]}],"b":[1,2]}\n'
# ---
# name: test_dump_int[json]
  b'1234'
# ---
# name: test_dump_int[jsonl]
  b'1234\n'
# ---
# name: test_dump_str[json]
  b'"just an ordinary\\n string"'
# ---
# name: test_dump_str[jsonl]
  b'"just an ordinary\\n string"\n'
# ---
# name: test_dump_to_file[tests/data/gen/api_array.json.gz]
  b'[{"a":"a1"},{"b":["b1","b2"]},{"c":{"c1":"c11"}}]'
# ---
# name: test_dump_to_file[tests/data/gen/api_array.json]
  b'[{"a":"a1"},{"b":["b1","b2"]},{"c":{"c1":"c11"}}]'
# ---
# name: test_dump_to_file[tests/data/gen/api_array.jsonl.gz]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file[tests/data/gen/api_array.jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file[tests/data/gen/api_array.ndjson.gz]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file[tests/data/gen/api_array.ndjson]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file_fails_without_file_name
  'You must give a FILENAME or USE_STDOUT_AS_DEFAULT=`True` or USE_STDERR_AS_DEFAULT=`True`'
# ---
# name: test_dump_to_file_stderr[None]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file_stderr[json]
  b'[{"a":"a1"},{"b":["b1","b2"]},{"c":{"c1":"c11"}}]'
# ---
# name: test_dump_to_file_stderr[jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file_stdout[None]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_dump_to_file_stdout[json]
  b'[{"a":"a1"},{"b":["b1","b2"]},{"c":{"c1":"c11"}}]'
# ---
# name: test_dump_to_file_stdout[jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_load_from_file_fails_without_file_name
  'You must give a FILENAME or USE_STDIN_AS_DEFAULT=`True`'
# ---
# name: test_sink[.json]
  b'[\n{"a":"a1"},\n{"b":["b1","b2"]},\n{"c":{"c1":"c11"}}\n]'
# ---
# name: test_sink[.jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink[.ndjson]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink_dict[.json]
  b'{"a":"b"}'
# ---
# name: test_sink_dict[.jsonl]
  b'{"a":"b"}\n'
# ---
# name: test_sink_dict[.ndjson]
  b'{"a":"b"}\n'
# ---
# name: test_sink_from_file[tests/data/gen/api_array.json.gz]
  b'[\n{"a":"a1"},\n{"b":["b1","b2"]},\n{"c":{"c1":"c11"}}\n]'
# ---
# name: test_sink_from_file[tests/data/gen/api_array.json]
  b'[\n{"a":"a1"},\n{"b":["b1","b2"]},\n{"c":{"c1":"c11"}}\n]'
# ---
# name: test_sink_from_file[tests/data/gen/api_array.jsonl.gz]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink_from_file[tests/data/gen/api_array.jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink_from_file[tests/data/gen/api_array.ndjson.gz]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink_from_file[tests/data/gen/api_array.ndjson]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink_from_file_fails_without_file_name
  'You must give a FILENAME or USE_STDOUT_AS_DEFAULT=`True` or USE_STDERR_AS_DEFAULT=`True`'
# ---
# name: test_sink_from_file_stdout[None]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
# name: test_sink_from_file_stdout[json]
  b'[\n{"a":"a1"},\n{"b":["b1","b2"]},\n{"c":{"c1":"c11"}}\n]'
# ---
# name: test_sink_from_file_stdout[jsonl]
  b'{"a":"a1"}\n{"b":["b1","b2"]}\n{"c":{"c1":"c11"}}\n'
# ---
